src/renderer/base/RenderSettings.cpp

```diff
@@ -0,0 +1,289 @@
+    if (mode == Mode::BlinkAllowed && !enabled)
+    {
+        _blinkShouldBeFaint = false;
+    }
+}
+
+// Routine Description:
+// - Retrieves the specified render mode.
+// Arguments:
+// - mode - The render mode to query.
+// Return Value:
+// - True if the mode is enabled. False if disabled.
+bool RenderSettings::GetRenderMode(const Mode mode) const noexcept
+{
+    return _renderMode.test(mode);
+}
+
+// Routine Description:
+// - Returns a reference to the active color table array.
+const std::array<COLORREF, TextColor::TABLE_SIZE>& RenderSettings::GetColorTable() const noexcept
+{
+    return _colorTable;
+}
+
+// Routine Description:
+// - Resets the first 16 color table entries with default values.
+void RenderSettings::ResetColorTable() noexcept
+{
+    InitializeColorTable({ _colorTable.data(), 16 });
+}
+
+// Routine Description:
+// - Creates the adjusted color array, which contains the possible foreground colors,
+//   adjusted for perceivability
+// - The adjusted color array is 2-d, and effectively maps a background and foreground
+//   color pair to the adjusted foreground for that color pair
+void RenderSettings::MakeAdjustedColorArray() noexcept
+{
+    // The color table has 16 colors, but the adjusted color table needs to be 18
+    // to include the default background and default foreground colors
+    std::array<COLORREF, 18> colorTableWithDefaults;
+    std::copy_n(std::begin(_colorTable), 16, std::begin(colorTableWithDefaults));
+    colorTableWithDefaults[AdjustedFgIndex] = GetColorAlias(ColorAlias::DefaultForeground);
+    colorTableWithDefaults[AdjustedBgIndex] = GetColorAlias(ColorAlias::DefaultBackground);
+
+    for (auto fgIndex = 0; fgIndex < 18; ++fgIndex)
+    {
+        const auto fg = til::at(colorTableWithDefaults, fgIndex);
+        for (auto bgIndex = 0; bgIndex < 18; ++bgIndex)
+        {
+            if (fgIndex == bgIndex)
+            {
+                _adjustedForegroundColors[bgIndex][fgIndex] = fg;
+            }
+            else
+            {
+                const auto bg = til::at(colorTableWithDefaults, bgIndex);
+                _adjustedForegroundColors[bgIndex][fgIndex] = ColorFix::GetPerceivableColor(fg, bg);
+            }
+        }
+    }
+}
+
+// Routine Description:
+// - Updates the given index in the color table to a new value.
+// Arguments:
+// - tableIndex - The index of the color to update.
+// - color - The new COLORREF to use as that color table value.
+void RenderSettings::SetColorTableEntry(const size_t tableIndex, const COLORREF color)
+{
+    _colorTable.at(tableIndex) = color;
+}
+
+// Routine Description:
+// - Retrieves the value in the color table at the specified index.
+// Arguments:
+// - tableIndex - The index of the color to retrieve.
+// Return Value:
+// - The COLORREF value for the color at that index in the table.
+COLORREF RenderSettings::GetColorTableEntry(const size_t tableIndex) const
+{
+    return _colorTable.at(tableIndex);

```

Comment:
esI'd suggest to use `gsl::at` here and for the setter above and to make the two functions noexcept.
I think the possibility of tableIndex being out of bounds is even rarer than "exceptional", which means it should be fine if we just abort() for invalid indices. But that way we can mark more functions as noexcept and drop try/catches.

Reply:
We definitely can't do this for `SetColorTableEntry` as it's currently used, and while I think it's probably safe for `GetColorTableEntry`, I suspect it won't be so in the future. If there's a desperate need to do this for performance reasons, it's possible we could make these safe, but that's not a change I'd want to attempt in this PR.

Reply:
Why can't this be done for `SetColorTableEntry` and why would it break `GetColorTableEntry` in the future?